src/llvm_api.cpp

```diff
@@ -0,0 +1,134 @@
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(orc::JITDylib, LLVMOrcJITDylibRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(orc::ExecutionSession, LLVMOrcExecutionSessionRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(orc::OrcV2CAPIHelper::PoolEntry,
+                                   LLVMOrcSymbolStringPoolEntryRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(orc::IRCompileLayer, LLVMOrcIRCompileLayerRef)
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(orc::MaterializationResponsibility,
+                                   LLVMOrcMaterializationResponsibilityRef)
+extern "C" {
+
+JL_DLLEXPORT_CODEGEN JuliaOJITRef LLVMExtraJLJITGetJuliaOJIT_impl(void)
+    JL_NOTSAFEPOINT
+{
+    return wrap(jl_ExecutionEngine);
+}
+
+JL_DLLEXPORT_CODEGEN LLVMOrcExecutionSessionRef
+LLVMExtraJLJITGetLLVMOrcExecutionSession_impl(JuliaOJITRef JIT) JL_NOTSAFEPOINT
+{
+    return wrap(&unwrap(JIT)->getExecutionSession());
+}
+
+JL_DLLEXPORT_CODEGEN LLVMOrcJITDylibRef
+LLVMExtraJLJITGetExternalJITDylib_impl(JuliaOJITRef JIT) JL_NOTSAFEPOINT
+{
+    return wrap(&unwrap(JIT)->getExternalJITDylib());
+}
+
+JL_DLLEXPORT_CODEGEN LLVMErrorRef LLVMExtraJLJITAddObjectFile_impl(
+    JuliaOJITRef JIT, LLVMOrcJITDylibRef JD, LLVMMemoryBufferRef ObjBuffer) JL_NOTSAFEPOINT
+{
+    return wrap(unwrap(JIT)->addObjectFile(
+        *unwrap(JD), std::unique_ptr<MemoryBuffer>(unwrap(ObjBuffer))));
+}
+
+JL_DLLEXPORT_CODEGEN LLVMErrorRef LLVMExtraJLJITAddLLVMIRModule_impl(
+    JuliaOJITRef JIT, LLVMOrcJITDylibRef JD, LLVMOrcThreadSafeModuleRef TSM) JL_NOTSAFEPOINT
+{
+    std::unique_ptr<orc::ThreadSafeModule> TmpTSM(unwrap(TSM));
+    return wrap(unwrap(JIT)->addExternalModule(*unwrap(JD), std::move(*TmpTSM)));
+}
+
+JL_DLLEXPORT_CODEGEN LLVMErrorRef
+LLVMExtraJLJITLookup_impl(JuliaOJITRef JIT, LLVMOrcExecutorAddress *Result,
+                                   const char *Name, int ExternalJDOnly) JL_NOTSAFEPOINT
+{
+    auto Sym = unwrap(JIT)->findExternalJDSymbol(Name, ExternalJDOnly);
+    if (Sym) {
+        auto addr = Sym->getAddress();
+        *Result = orc::ExecutorAddr(addr).getValue();
+        return LLVMErrorSuccess;
+    }
+    else {
+        *Result = 0;
+        return wrap(Sym.takeError());
+    }
+}
+
+JL_DLLEXPORT_CODEGEN LLVMOrcSymbolStringPoolEntryRef
+LLVMExtraJLJITMangleAndIntern_impl(JuliaOJITRef JIT,
+                                            const char *Name) JL_NOTSAFEPOINT
+{
+    return wrap(orc::OrcV2CAPIHelper::moveFromSymbolStringPtr(unwrap(JIT)->mangle(Name)));
+}
+
+JL_DLLEXPORT_CODEGEN const char *
+LLVMExtraJLJITGetTripleString_impl(JuliaOJITRef JIT)
+{
+    return unwrap(JIT)->getTargetTriple().str().c_str();
+}
+
+JL_DLLEXPORT_CODEGEN const char
+LLVMExtraJLJITGetGlobalPrefix_impl(JuliaOJITRef JIT)
+{
+    return unwrap(JIT)->getDataLayout().getGlobalPrefix();
+}
+
+JL_DLLEXPORT_CODEGEN const char *
+LLVMExtraJLJITGetDataLayoutString_impl(JuliaOJITRef JIT)
+{
+    return unwrap(JIT)->getDataLayout().getStringRepresentation().c_str();
```

Comment:
The C API supports DataLayout objects, so just return it directly?

Reply:
Oh, I didn't know that, because the other functions (GetDataLayout,SetDatalayout etc) that move it around all use C strings

Reply:
Ah, I see that the existing APIs use strings too (`LLVMOrcLLJITGetDataLayoutStr`), so I'm OK keeping it as strings for consistency.